### PR TITLE
ESP32 classic fixes and a new board

### PR DESCRIPTION
**Pull request by davidbuzz**

### CODE REVIEW

File: libraries/AP_InertialSensor/AP_InertialSensor_Invensense.cpp [hunk 1]
```diff
@@ -587,7 +587,9 @@ bool AP_InertialSensor_Invensense::_accumulate(uint8_t *samples, uint8_t n_sampl
                 return false;
             } else {
                 if (!hal.scheduler->in_expected_delay()) {
+                    #ifndef HAL_ESP32_BOARD_NAME
```
Comment (nkruzan) [inline libraries/AP_InertialSensor/AP_InertialSensor_Invensense.cpp:590]: is there any particular reason to require additional setting in hwdef for these to go away?
could these be HAL_BOARD_NAME == "ESP32" instead?
really we are getting kind of convoluted 
HAL_BOARD_NAME
HAL_ESP32_BOARD_NAME
HAL_BOARD_TYPE
HAL_BOARD_SUBTYPE
used in various places to answer the "IS_THIS_ESP32" question

/rant

**Review by nkruzan — CHANGES_REQUESTED** [review]

File: libraries/AP_InertialSensor/AP_InertialSensor_Invensense.cpp (hunk 1 above)
Comment (khancyr) [reply libraries/AP_InertialSensor/AP_InertialSensor_Invensense.cpp:590]: probably just add a case in the debug() function definition on start to nuke it on esp. That will be simpler that adding ifndef everywhere

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_LORD.cpp [hunk 2]
```diff
@@ -13,8 +13,9 @@
 /*
   suppport for LORD Microstrain CX5/GX5-45 serially connected AHRS Systems
  */
-
+#ifndef ALLOW_DOUBLE_MATH_FUNCTIONS
```
[the remark this answers is not in the document]
Comment (davidbuzz) [reply libraries/AP_ExternalAHRS/AP_ExternalAHRS_LORD.cpp:16]: fixed in latest version pf pr

File: libraries/AP_HAL/board/esp32.h [hunk 3]
```diff
@@ -61,3 +68,54 @@
...
+// turn off all the compasses.. except gy-91's MPU9250 
+#define AP_COMPASS_DIAGONALS_ENABLED 0
+#define AP_COMPASS_BACKEND_DEFAULT_ENABLED 0
+#define AP_COMPASS_EXTERNALAHRS_ENABLED 0
+#define AP_COMPASS_MSP_ENABLED 0
+#define AP_COMPASS_SITL_ENABLED 0
+#define AP_COMPASS_UAVCAN_ENABLED 0
+#define AP_COMPASS_I2C_BACKEND_DEFAULT_ENABLED 0
+#define AP_COMPASS_AK09916_ENABLED 0
+#define AP_COMPASS_AK8963_ENABLED 0
```
[the remark this answers is not in the document]
Comment (davidbuzz) [reply libraries/AP_HAL/board/esp32.h:97]: fixed in latest version pf pr

File: libraries/AP_InertialSensor/AP_InertialSensor_Invensense.cpp (hunk 1 above)
Comment (davidbuzz) [reply libraries/AP_InertialSensor/AP_InertialSensor_Invensense.cpp]: fixed in latest version pf pr - i took @khancyr's suggestion on this one, and made the debug() call esp32 aware

**Comment by davidbuzz** [comment]
added dev-call-topic as there's a few things outside of AP_HAL_ESP32 that might offend some people , and I'd like to see if its ok or not.

File: libraries/AP_HAL/board/esp32.h [hunk 4]
```diff
@@ -61,3 +67,99 @@
 #define CONFIG_ESP32_WIFI_TX_BA_WIN 0
 #define CONFIG_ESP32_WIFI_RX_BA_WIN 0
 
+
+// turn off all the compasses by default.. 
+#ifndef AP_COMPASS_DIAGONALS_ENABLED
+#define AP_COMPASS_DIAGONALS_ENABLED 0
+#endif
+#ifndef AP_COMPASS_BACKEND_DEFAULT_ENABLED
```
Comment (peterbarker) [inline libraries/AP_HAL/board/esp32.h:75]: This should be sufficient to turn all of the backends off that you don't want - you don't need to specify the others.

File: libraries/AP_HAL/board/esp32.h [hunk 5]
```diff
@@ -61,3 +67,99 @@
 #define CONFIG_ESP32_WIFI_TX_BA_WIN 0
 #define CONFIG_ESP32_WIFI_RX_BA_WIN 0
 
+
+// turn off all the compasses by default.. 
+#ifndef AP_COMPASS_DIAGONALS_ENABLED
```
Comment (peterbarker) [inline libraries/AP_HAL/board/esp32.h:72]: This isn't a backend!

File: libraries/AP_HAL_ESP32/boards/esp32buzz.h [hunk 6]
```diff
@@ -65,6 +65,12 @@
 // BARO probing:
 //#define HAL_BARO_PROBE_LIST PROBE_BARO_I2C(BMP280, 0, 0x77)
 
+// no airspeed sensor
+#define AP_AIRSPEED_MS4525_ENABLED 0
+#define AP_AIRSPEED_ENABLED 0
+#define AP_AIRSPEED_ANALOG_ENABLED 0
+#define AP_AIRSPEED_BACKEND_DEFAULT_ENABLED 0
```
Comment (peterbarker) [inline libraries/AP_HAL_ESP32/boards/esp32buzz.h:72]: You shouldn't need to set the backends individually - just the backend_default parameter

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_VectorNav.cpp [hunk 7]
```diff
@@ -16,7 +16,9 @@
   suppport for serial connected AHRS systems
  */
 
+#ifndef ALLOW_DOUBLE_MATH_FUNCTIONS
```
Comment (tridge) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_VectorNav.cpp:19]: change in class esp32 to not set to 1

**Review by peterbarker — COMMENTED** [review]
Your HAL and board defs, so you can do what you like - but you're borrowing trouble by enumerating the compasses etc as you are :-)

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_VectorNav.cpp (hunk 7 above)
Comment (tridge) [reply libraries/AP_ExternalAHRS/AP_ExternalAHRS_VectorNav.cpp:19]: should set HAL_HAVE_HARDWARE_DOUBLE=1 instead

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by davidbuzz** [comment]
... confirmed... I have just finished testing this PR  and it compiling/flashing/booting plane, copter and rover on classic esp32  with my 'esp32buzz' board, which is a big step up from without this PR, where they either don't compile, or crash on startup.

**Comment by davidbuzz** [comment]
[sidenote, but @tridge  asked for it ] I'd like to note that is PR also includes a cpuinfo output from the esp32.  I've also done up a comparison spreadsheet with the results of all the boards in that folder...
https://docs.google.com/spreadsheets/d/1Tz1tIc8ISPQh6fYKJ1dCplgqioLBaCYkUpmRe_glmqo/edit?usp=sharing

File: Tools/CPUInfo/output-esp32-classic.txt [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+isub32      0.0614 usec/call
+imul32      0.0612 usec/call
+idiv32      0.0654 usec/call
+iadd64      0.1232 usec/call
+isub64      0.1232 usec/call
+imul64      0.1712 usec/call
+idiv64      0.3848 usec/call
+memcpy128   0.5094 usec/call
+memset128   0.3917 usec/call
+delay(1)   164.9920 usec/call
```
Comment (WickedShell) [inline Tools/CPUInfo/output-esp32-classic.txt:93]: Isn't this indicative of a bug? This is seriously off of the expected result.

Comment (davidbuzz) [reply Tools/CPUInfo/output-esp32-classic.txt:93]: possibly, but not one in this PR.  :-)

File: ArduCopter/autoyaw.cpp [hunk 9]
```diff
@@ -110,10 +110,10 @@ void Mode::AutoYaw::set_fixed_yaw(float angle_deg, float turn_rate_ds, int8_t di
 
     // calculate final angle as relative to vehicle heading or absolute
     if (relative_angle) {
-        _fixed_yaw_offset_cd = angle_deg * 100.0 * (direction >= 0 ? 1.0 : -1.0);
+        _fixed_yaw_offset_cd = angle_deg * 100.0f * (direction >= 0 ? 1.0 : -1.0);
```
Comment (tridge) [inline ArduCopter/autoyaw.cpp:113]: should use -fsingle-precision-constant in CPP flags

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ESP32/targets/esp-idf/project_description.json [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    },
+    "build_components" : [ "bootloader", "bootloader_support", "efuse", "esp32", "esp_common", "esp_rom", "esptool_py", "log", "main", "micro-ecc", "partition_table", "soc", "spi_flash", "xtensa", "" ],
+    "build_component_paths" : [ "/home/buzz/ardupilot/modules/esp_idf/components/bootloader", "/home/buzz/ardupilot/modules/esp_idf/components/bootloader_support", "/home/buzz/ardupilot/modules/esp_idf/components/efuse", "/home/buzz/ardupilot/modules/esp_idf/components/esp32", "/home/buzz/ardupilot/modules/esp_idf/components/esp_common", "/home/buzz/ardupilot/modules/esp_idf/components/esp_rom", "/home/buzz/ardupilot/modules/esp_idf/components/esptool_py", "/home/buzz/ardupilot/modules/esp_idf/components/log", "/home/buzz/ardupilot/modules/esp_idf/components/bootloader/subproject/main", "/home/buzz/ardupilot/modules/esp_idf/components/bootloader/subproject/components/micro-ecc", "/home/buzz/ardupilot/modules/esp_idf/components/partition_table", "/home/buzz/ardupilot/modules/esp_idf/components/soc", "/home/buzz/ardupilot/modules/esp_idf/components/spi_flash", "/home/buzz/ardupilot/modules/esp_idf/components/xtensa", "" ]
```
Comment (tridge) [inline libraries/AP_HAL_ESP32/targets/esp-idf/project_description.json:19]: only users called buzz can build? :-)

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduCopter/autoyaw.cpp (hunk 9 above)
Comment (davidbuzz) [reply ArduCopter/autoyaw.cpp]: done

File: libraries/AP_HAL_ESP32/targets/esp-idf/project_description.json (hunk 10 above)
Comment (davidbuzz) [reply libraries/AP_HAL_ESP32/targets/esp-idf/project_description.json]: I've removed that file/commit, it was a template for something still a bit wip.

File: libraries/AP_Logger/AP_Logger.cpp [hunk 11]
```diff
@@ -36,7 +36,11 @@ extern const AP_HAL::HAL& hal;
 #endif
 
 #ifndef HAL_LOGGING_STACK_SIZE
-#define HAL_LOGGING_STACK_SIZE 1580
+    #if CONFIG_HAL_BOARD == HAL_BOARD_ESP32
```
Comment (peterbarker) [inline libraries/AP_Logger/AP_Logger.cpp:39]: Why not define this in the AP_HAL_Boards.h file - or in the ESP32 include file included from there?

That's why the #ifdef exists here!

Comment (tridge) [inline libraries/AP_Logger/AP_Logger.cpp:39]: should be in ESP32 board.h

File: libraries/GCS_MAVLink/GCS_FTP.cpp [hunk 12]
```diff
@@ -47,9 +47,13 @@ bool GCS_MAVLINK::ftp_init(void) {
     if (ftp.requests == nullptr || ftp.requests->get_size() == 0) {
         goto failed;
     }
-
+    #if CONFIG_HAL_BOARD == HAL_BOARD_ESP32
```
Comment (tridge) [inline libraries/GCS_MAVLink/GCS_FTP.cpp:50]: should just be an ifndef, then add to board.h

**Review by tridge — CHANGES_REQUESTED** [review]
commits should have subsystem prefix
I think you will find that adding a base amount in thread_create() for ESP32 interrupt handler works better than fiddling each case

File: libraries/AP_Logger/AP_Logger.cpp (hunk 11 above)
Comment (davidbuzz) [reply libraries/AP_Logger/AP_Logger.cpp]: I was unaware that chibios had n internal "extra bit" that it adds to each thread created through Scheduler::thread_create ... i've done equivalent in the esp32 HAL now.

Comment (davidbuzz) [reply libraries/AP_Logger/AP_Logger.cpp]: I was unaware that chibios had an internal "extra bit" that it adds to each thread created through Scheduler::thread_create ... i've done equivalent in the esp32 HAL now.

File: libraries/GCS_MAVLink/GCS_FTP.cpp (hunk 12 above)
Comment (davidbuzz) [reply libraries/GCS_MAVLink/GCS_FTP.cpp]: I was unaware that chibios had an internal "extra bit" that it adds to each thread created through Scheduler::thread_create ... i've done equivalent in the esp32 HAL now.

**Review by tridge — APPROVED** [review]

**Pull request merged**